File: 0012-integer-to-roman/0012-integer-to-roman.py
class Solution:
    def intToRoman(self, num: int) -> str:
        symbols = {
            1000: 'M',
            500: 'D',
            100: 'C',
            50: 'L',
            10: 'X',
            5: 'V',
            1: 'I'
        }

        unit_values = [1000, 100, 10, 1]        

        answer = ""
        for val in unit_values:
            quot = num//val
            num = num%val

            if quot == 4 or quot == 9:
                if val >= 1000:
                    pass
                elif val >= 100:
                    answer += "CM" if quot == 9 else "CD"
                elif val >= 10:
                    answer += "XC" if quot == 9 else "XL"
                else:
                    answer += "IX" if quot == 9 else "IV"
            else:
                if quot == 0:
                    pass
                elif quot < 4:
                    for i in range(quot):
                        answer += symbols[val]
                elif quot == 5:
                    answer += symbols[quot*val]
                elif quot < 9:
                    answer += symbols[5*val]
                    for i in range(quot - 5):
                        answer += symbols[val]

        
        return answer

File: 0012-integer-to-roman/0012-integer-to-roman.java
class Solution {
    public String intToRoman(int num) {
        StringBuilder sb = new StringBuilder("");
        int temp = num;

        int divider = 1000;
        while (divider > 0){
            int quot = temp / divider;
            temp = temp % divider;

            if (quot == 0){
                divider = divider/10;
                continue;
            }

            sb.append(getString(quot, divider));
            divider /= 10;
        }

        return sb.toString();
    }

    private String getString(int quot, int divider){
        String digit_1 = "", digit_5 = "", digit_10 = "";
        switch (divider){
            case 1:
                digit_1 = "I";
                digit_5 = "V";
                digit_10 = "X";
                break;
            case 10:
                digit_1 = "X";
                digit_5 = "L";
                digit_10 = "C";
                break;

            case 100:
                digit_1 = "C";
                digit_5 = "D";
                digit_10 = "M";
                break;

            case 1000:
                digit_1 = "M";
                break;
            }
        String roman = "";
        switch (quot){
            case 4:
                roman += digit_1 + digit_5;
                break;
            case 5:
                roman += digit_5;
                break;
            case 9:
                roman += digit_1 + digit_10;
                break;
            default:
                if(quot < 5){
                    for (int i = 0; i < quot; i++){
                        roman += digit_1;
                    }
                } else if (quot < 9){
                    roman += digit_5;
                    for (int i = 0; i < quot-5; i++){
                        roman += digit_1;
                    }   
                }
        }
        return roman;
    }
}